keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+
+  *Limitation*: Currently, it’s impossible to specify the desired
+  device properties that are required for reconfiguring devices.
+  For the FPGA example, a file containing the desired configuration
+  of the FPGA has to be referenced.
+
+- *Device cleanup*: When my workload is finished, I would like to have
+  a mechanism for cleanup of the device, that will ensure that device
+  does not contain traces/parameters/data from previous workloads and
+  appropriate power state/shutdown. For example, an FPGA might have
+  to be reset because its configuration for the workload was
+  confidential.
+
+  *Limitation*: Post-stop actions are not supported.
+
+- *Partial allocation*: When deploying a container I’d like to be able
+  to use part of the shareable device inside a container and other
+  containers should be able to use other free resources on the same
+  device.
+
+  *Limitation*: Current implementation of the device plugin doesn’t
+  allow one to allocate part of the device because parameters are too limited
+  and Kubernetes doesn't have enough information about the extended
+  resources on a node to decide whether they can be shared.
+
+- *Optional allocation*: When deploying a workload I’d like to specify
+  soft(optional) device requirements. If a device exists and it’s
+  allocatable it will be allocated. If not - the workload will be run on
+  a node without a device. GPU and crypto-offload engines are
+  examples of this kind of device. If they’re not available, workloads
+  can still run by falling back to using only the CPU for the same
+  task.
+
+  *Limitation*: Optional allocation is supported neither by the device
+  plugins nor by current Pod resource declaration.
+
+- *Support Over the Fabric devices*: When deploying a container, I’d
+  like to utilize devices available over the Fabric (network, special
+  links, etc).
+
+  *Limitation*: Device Plugins framework supports only local devices.
+
+Several other limitations are addressed by
+[CDI](https://github.com/container-orchestrated-devices/container-device-interface/).
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+* More flexibility:
+  * Arbitrary parameters for resource requests
+  * Network-attached resources
+  * Arbitrary, resource-specific setup and cleanup actions
+  * Custom matching of resource requests with available resources,
+    including handling of optional resource requests
+* User-friendly API for describing resource requests
+* Allow resource management cluster add-ons that can be developed and deployed
+  completely free of requiring to re-build any core Kubernetes component
+  and that are independent of specific container runtimes.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+* Replace the device manager API. For resources that fit into its model
```

Comment:
The new API can express everything that the old one can and more, so from a semantic point of view it is a replacement. We can state that as a goal (added):

> * Rich enough semantic so that all current device plugins could be implemented based on dynamic resource allocation

But because it is more flexible, it cannot do some of the optimizations that were possible with the old one (primarily assigning resources to pods entirely in the scheduler). There's also the question whether we want to force the ecosystem to use something else. So overall I think making "replace device plugin API"  a non-goal is the right choice.

If this new approach turns out to be so successful that everyone uses it, then we can still decide to deprecate the device plugin API. However, then it would have to stay in beta (current state) until we know that.

Apropos: should the KEP use "Device Plugin API" or "device plugin API"? https://github.com/kubernetes/design-proposals-archive/blob/main/resource-management/device-plugin.md itself seems to handle this inconsistently.

Lower case seems to be more common, so I've switched to that throughout the KEP.

Reply:
Device plugin is not a formal name, it's just a description. :)

Agree that "replace it" is not a goal, but "ensure that this could serve as a replacement is.  Thanks.